overlord/udevmon.go

```diff
@@ -0,0 +1,124 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package overlord
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/snapcore/snapd/logger"
+
+	"github.com/pilebones/go-udev/crawler"
+	"github.com/pilebones/go-udev/netlink"
+)
+
+var (
+	udevMon *UDevMonitor
+	once    sync.Once
+)
+
+type udevMonitor interface {
+	Run() error
+	Stop()
+}
+
+type UDevMonitor struct {
+	netlinkConn *netlink.UEventConn
+	stop        chan struct{}
+	monitorStop chan struct{}
+	crawlerStop chan struct{}
+}
+
+func NewUDevMonitor() (*UDevMonitor, error) {
+	var err error
+	// There can only ever be one instance of netlink monitor per process,
+	// use singleton pattern to force this as otherwise it's very painful
+	// for the unit tests that do not delete overlord instances, or
+	// instantiate more than one overlord.
+	once.Do(func() {
+		udevMon = &UDevMonitor{
+			stop:        make(chan struct{}),
+			netlinkConn: &netlink.UEventConn{},
+		}
+		err = udevMon.netlinkConn.Connect()
```

Comment:
I think that connecting here makes things a bit complicated. AFAIU `NewUdevMonitor()` should just return:
```
&UDevMonitor{
    stop: make(chan struct{}),
}
```
and that's it.

Then in `Run()` you establish the connection and so on. IMO this should also remove the problem with having to use once. You'll get multiple instances of UDevMonitor, but none of them are connected, so gc will clean that up at some point.

Reply:
Done, see earlier explanation too.